### PR TITLE
Bug 1710502: Implementing graceful shutdown.

### PR DESCRIPTION
**Pull request by ricardomaraschini**
Running webserver in a separated go routine and awaiting for SIGTERM.

### CODE REVIEW

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 1]
```diff
@@ -164,14 +166,25 @@ func Execute(configFile io.Reader) {
...
+			dcontext.GetLogger(ctx).Infof("listening on %s", srv.Addr)
+			errc <- srv.ListenAndServe()
+		} else {
+			dcontext.GetLogger(ctx).Infof("listening on %s, tls", srv.Addr)
+			errc <- srv.ListenAndServeTLS(dockerConfig.HTTP.TLS.Certificate, dockerConfig.HTTP.TLS.Key)
+		}
+	}()
+
+	<-sigs
+	if err := srv.Close(); err != nil {
```
Comment (dmage) [inline pkg/cmd/dockerregistry/dockerregistry.go]: You can use Shutdown for a graceful shutdown.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 2]
```diff
@@ -164,16 +166,36 @@ func Execute(configFile io.Reader) {
...
+	select {
+	case <-sigs:
+	case err := <-errc:
+		log.Fatal(err)
 	}
-	if err != nil {
+
+	// signal has been received, attempt to stop the server within the
+	// default termination grace period of 30 seconds.
+	ctx, cancel := context.WithTimeout(ctx, 29*time.Second)
```
Comment (dmage) [inline pkg/cmd/dockerregistry/dockerregistry.go]: > period of 30 seconds

> `29*time.Second`

Why are the numbers different?

Comment (ricardomaraschini) [reply pkg/cmd/dockerregistry/dockerregistry.go]: The grace period is 30 seconds, i was planning to give it 1 entire second between receiving the signal and calling Shutdown(), maybe i could update the comment, what do you think?

Comment (dmage) [reply pkg/cmd/dockerregistry/dockerregistry.go]: Yes, it worth documenting this difference so that the next reader won't be surprised.

Comment (ricardomaraschini) [reply pkg/cmd/dockerregistry/dockerregistry.go]: done.

**Comment by ricardomaraschini** [comment]
> /hold
> 
> We need an e2e for this, unfortunately the registry operator is the best place to do it. Test should do the following:
> 
> 1. Make sure a registry is available
> 2. Update the operator to scale the registry down to 0
> 3. Wait 30 seconds
> 4. Verify there are no registry pods.
> 
> I'll cancel the hold once the e2e PR is up.

Implemented(with a slightly different approach) on https://github.com/openshift/cluster-image-registry-operator/pull/372

**Pull request merged**